### PR TITLE
added structured 1_explore_babyweight

### PR DESCRIPTION
**Pull request by ryangillard**

### CODE REVIEW

File: courses/machine_learning/deepdive2/structured/labs/1_explore_babyweight.ipynb [hunk 1]
```diff
@@ -0,0 +1,345 @@
+{
```
Comment (danideter) [inline courses/machine_learning/deepdive2/structured/labs/1_explore_babyweight.ipynb]: <p>Line too long, &gt;80 characters.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb [hunk 2]
```diff
@@ -0,0 +1,536 @@
+{
```
Comment (danideter) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>The x axis here looks a little crammed. Can you please adjust the labels? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (danideter) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Missing comma after "In this lab"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

**Comment by danideter** [comment]
First lab looks good, Ryan. Minor comments

File: courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb (hunk 2 above)
Comment (danideter) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>This is a security risk and there should either be a warning to students, or should be done in a way that prevents SQL injection</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

File: courses/machine_learning/deepdive2/structured/labs/1_explore_babyweight.ipynb (hunk 1 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/labs/1_explore_babyweight.ipynb]: <p>Done</p>

File: courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb (hunk 2 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>These are learning how to do ML notebooks, we're not trying to protect against every form of SQL injection. This is just trying to get a simple concept across, it is up to the user to harden any code they want to use in the real world outside of the classroom.</p>

**Comment by ryangillard** [comment]
All comments responded to, thanks for the review!

File: courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb (hunk 2 above)
Comment (danideter) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: Then do the former and add a quick note and [a link](https://cloud.google.com/bigquery/docs/parameterized-queries). I've had multiple students flag to me that this is not industry best practice. Let's put them at ease that we know what we're doing :)

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: You can't parameterize column names it seems, those are not scalars, arrays of scalars, or structs of scalars. Therefore, I can't do the former to leave a note or link because the link's method doesn't work for column names. And I can't do the latter because it doesn't work.

From the documentation:
Parameters cannot be used as substitutes for identifiers, column names, table names, or other parts of the query.

Comment (danideter) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: Right, Bigquery doesn't have an easy solution, but that doesn't make this code safe.
What you could do is make a set of acceptable column names, and see if the passed variable is in the set, and mention this is a crude way of preventing SQL injection for instructional purposes. It should be, at most, three lines of code to be added.

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Added SQL Injection protection to code.</p>

**Comment by ryangillard** [comment]
<p>All comments resolved, thanks for the review!</p>

**Comment by danideter** [comment]
Much more professional! Great job Ryan, LGTM.
I'll leave the merging honors to you.

File: courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <ul><li> Something is missing here: "Use Cloud AI Platform Notebooks plot explorations" </li><li class="ql-indent-1">Maybe replace by "Learn how to explore a dataset by using plots in the CAIP notebooks"</li><li><strong>"#TODO</strong>&nbsp;in this student lab notebook" -&gt; "<strong>#TODO</strong>&nbsp;in this  notebook" (this is the solution not the lab, right?)</li><li>"-- try to complete this notebook first and then review the&nbsp;<a href="https://github.com/GoogleCloudPlatform/training-data-analyst/raw/867c2482916c774e183cbea40f3d30b64711aed3/courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb" target="_blank">solution notebook</a>." -&gt; maybe remove this part since it is the solution notebook</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>* nit: "My goal" -&gt; "The goal"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>I'd move the comment in a markdown cell above instead of having it in the code, and I'd rephrase it as follow:</p><p> </p><p>"Let's now create a SQL to retrieve the natality dataset after the year 2000:"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>I'd move the comment in a markdown cell above instead of having it in the code, and I'd rephrase it as follow:</p><p> </p><p>"We now examine the result of a BiqQuery call in a Pandas DataFrame: "</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <ul><li>Do we need to create the BigQuery client each time we issue a query? If not, just create one client at the beginning of the notebook and reuse it throughout the notebook.</li><li>Why are we calling these columns "valid"? From the query, it seems we are extracting all the columns from that table. If no real reason to call these columns "valid", let's rename the variables as follows: valid_columns_df -&gt; columns_df and valid_columns_set -&gt; columns_set</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>* "valid columns" -&gt; columns (see comment above)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>* valid_columns_set -&gt; columns_set (see comment above)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <ul><li>valid_columns_set -&gt; columns_set (see comment above)</li><li>Remove the code comment directly above the function (You already have a docstring explaining the same thing)</li><li>In the docstring: </li><li class="ql-indent-1">statstics -&gt; statistics</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Remove the code comment directly above the function, since you already have the docstring.</p><p> </p><p>In the docstring:</p><ul><li>statstics -&gt; statistics</li><li>valid_columns_set -&gt; columns_set (see comment above)</li></ul><p> </p><p>In the function body:</p><ul><li>valid_columns_set -&gt; columns_set (see comment above)</li><li>pep8: Break the following line (it's too long):</li></ul><p> </p><p><code>&nbsp;&nbsp;df.plot(x=column_name, y="num_babies", logy=logy, kind="bar", figsize=(12, 5))</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Move the code comment into a markdown cell, and make a full sentence with it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Move the code comment into a markdown cell, and make a full sentence with it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Move the code comment into a markdown cell, and make a full sentence with it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Move the code comment into a markdown cell, and make a full sentence with it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>nit: "I will" -&gt; "We will"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>nit: remove (it's just repeating the lab ojective)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

**Comment by BenoitDherin** [comment]
<p>Please replicate the fixes from the solution comments to the lab. </p>

File: courses/machine_learning/deepdive2/structured/labs/1_explore_babyweight.ipynb [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive2/structured/labs/1_explore_babyweight.ipynb]: <p>I'd add the following hint into the TODO:</p><p> </p><p>"TODO: Reusing the get_distinct_values function you just implemented,  create function that plots distinct value statistics from BigQuery"</p><p> </p><p>I'd also put the todo into its own markdown cell instead of a code comment.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/665/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

**Comment by BenoitDherin** [comment]
<p>The lab looks good to me. Only minor suggestions and typos. I'll let you merge, when you are done with the fixes. </p>

File: courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb (hunk 3 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <ul><li>Good suggestion, changed to create client just once.</li><li>It is the set of all valid columns that a user could try as part of the SQL Injection hacking attack prevention. I have valid in the name to reinforce that these are the only valid columns a user can choose from to make clear the SQL Injection prevention.</li></ul>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Same as above</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Same as above.</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Just a nice reminder of what they just did instead of scrolling all of the way to the top. :)</p>

File: courses/machine_learning/deepdive2/structured/labs/1_explore_babyweight.ipynb (hunk 4 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/labs/1_explore_babyweight.ipynb]: <p>Done</p>

File: courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb (hunk 3 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/1_explore_babyweight.ipynb]: <p>Done</p>

**Comment by ryangillard** [comment]
<p>Thanks for the review Benoit! I made all of the changes and will commit to master since this was already merged after Danielle's completed review.</p>
